README.md
# Projet intégration SOC
Projet intégration SOC 3a 2017

# Application androïd
Faite un build et un launch ensuite vous pouvez l'utilisé
  https://github.com/florian-texier/Projet_integration_SOC/tree/master/Application

# Base de donner
Pour installer la base de donnée couchdb, rentré cette ligne de commande dans un terminal.
```
docker pull gauthier66/couchdb
```
Ensuite créer le ficher docker-compose.yml et y incorporer le yml suivant:
```
version: "3"
     services:
couchdb:
       	    container_name: couchdb 
 		    image: klaemo/couchdb:latest
    networks: [interne]
    labels:
      - traefik.backend=couchdb
      - traefik.frontend.rule=Host:couchdb.mignolet.fr
      - traefik.port=5984
networks:
  interne:
    external:
      name: interne

```
Ensuite lancer le docker
```
docker-compose up -d
```

# Rancher
Pour installer rancher suiver le tutoriel présent sur le site.
  http://rancher.com/

# Création du swarm avec rancher
## Installer le service router
```
docker pull gauthier66/routerint
```
Vous pouvez consulter les sources du python à cette adresse:
  https://github.com/mignolet/scavengerRouter.git

Ensuite créer le ficher docker-compose.yml et y incorporer le yml suivant:
```
version: '2'
			services:
  				router:
   				   image: gauthier66/routerint
    				environment:
  				    LOGINCOUCHDB: admin
  				    PWDCOUCHDB: adminsoc
  				  ports:
 				   - 4000:5000/tcp
```
Ensuite lancer le docker
```
docker-compose up -d
```
## Installer le service instance equipe
```
docker pull gauthier66/instanceequipe_web
```
Vous pouvez consulter les sources du python à cette adresse:
  https://github.com/mignolet/instanceEquipe.git

Ensuite créer le ficher docker-compose.yml et y incorporer le yml suivant:
```
version: '2'
			services:
  			 web:
  			  image: gauthier66/instanceequipe_web
```
Ensuite lancer le docker
```
docker-compose up -d
```

# Topic MQTTCLOUD
Ce topic va permettre le retour d'affichage sur la raspberry pi.
Vous pouvez retrouver le code pythona cette adresse:
  https://github.com/florian-texier/Projet_integration_SOC/tree/master/pythonRasp


HISTORY.md
# HISTORY
## 06/11/2017

Test des différents programme de démonstration.
Reflexion sur les différentes technologie utilisé.
Ebauche de l'architecture.
Réalisation du schéma de l'application.

## 07/11/2017

Réalisation de la première page de l'application.
Documentation et réalisation des chemins du routeur.
Création base de donnée.

## 08/11/2017

Réalisation de la deuxième page de l'application.
Développement du router docker

## 09/11/2017

Réalisation de la troisième page de l'application.
Replica docker suite a la perder de connection avec ovh.
Développement serveur python sur raspberry pi.

## 10/11/2017

Ajout de la caméro et de la géolocalisation.
Développement du router.

## 13/11/2017

Connection de l'application avec le webservice et correction de bug.
Développement router + instance utilisateur.

## 14/11/2017

Ajouts du selectionneur de fichier et affichage de l'image sélectionner.
Développement instance utilisateur.
Gestion de la scalabilité.

## 15/11/2017

Ajouts de l'envoie de la photo sur le serveur.

## 16/11/2017

Ajouts du logo et de la condition pour l'envoie de photo



Application/app/src/test/java/imerir/scavengerhunt/ExampleUnitTest.java
package imerir.scavengerhunt;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

routeint/serveur_flask.py
from json_responses import json_response, json_data, json_error
from flask import Flask, request
from flask_cors import CORS
import requests

app = Flask(__name__)
app.debug = True


@app.route('/')
def hello():
    return json_response('Hello World, from router api !')

#id inscription beacon
@app.route('/beaconInscript', methods=['GET'])
def beaconInscript():
    print("beaconInscript")
    r = requests.get('https://couchdb.mignolet.fr/beacondb/_design/type/_view/join?reduce=false')
    return json_response(r.json(),r.status_code)

#inscription team
@app.route('/inscript', methods=['POST'])
def inscript():
    #android id
    print("inscription team")
    id = request.get_json(force=True)
    jsonData = id["device_id"]
    r = requests.put('https://admin:adminsoc@couchdb.mignolet.fr/teambd/"002"')
    return json_response(r.json(), r.status_code)

#id beacon send picture
@app.route('/beaconSendPicture', methods=['GET'])
def beaconSendPicture():
    print("beaconInscript")
    r = requests.get('https://couchdb.mignolet.fr/beacondb/_design/type/_view/depot?reduce=false')
    return json_response(r.json(), r.status_code)

#liste objet search
@app.route('/newListe' , methods=['GET'])
def newsListe():
    print("newsListe")
    r = requests.get('https://couchdb.mignolet.fr/objetdb/_design/_all_dbs/_view/random?limit=10&reduce=false')
    return json_response(r.json(), r.status_code)

#liste objet search
@app.route('/listObjet' , methods=['GET'])
def ListObjet():
    print("item liste")
    r = requests.get('')
    return json_response(r.json(), r.status_code)

#inscrption instance scale
@app.route('/inscriptionInstance')
def inscrptionInsrance():
    return 'Hello World!'

#link instance for team
@app.route('/linkInstance')
def linkInstance():
    return 'Hello World!'


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000)



pythonRasp/mainFlaskRasp.py
from flask import Flask
from flask import request
from time import sleep
import datetime
from sense_hat import SenseHat

#init SenseHAT
sense = SenseHat()
sense.low_light = True
sense.set_rotation(180)
sense.clear(0,0,0)

#color led 
GREEN = [94,255,51]
RED = [255,0,0]
BLUE = [14,51,240]
O = [0, 0, 0] # white
YL = [200,0,255]

valider = [
O,O,O,O,O,O,O,O,
O,O,O,O,O,O,O,O,
O,O,O,O,O,O,O,O,
O,O,O,O,O,O,O,BLUE,
O,O,O,O,O,O,BLUE,O,
O,BLUE,O,O,O,BLUE,O,O,
O,O,BLUE,O,BLUE,O,O,O,
O,O,O,BLUE,O,O,O,O
]

fail = [
RED,O,O,O,O,O,O,RED,
O,RED,O,O,O,O,RED,O,
O,O,RED,O,O,RED,O,O,
O,O,O,RED,RED,O,O,O,
O,O,O,RED,RED,O,O,O,
O,O,RED,O,O,RED,O,O,
O,RED,O,O,O,O,RED,O,
RED,O,O,O,O,O,O,RED
]

t = [
O,O,BLUE,O,O,O,O,O,
O,BLUE,BLUE,BLUE,O,O,O,O,
BLUE,O,BLUE,O,BLUE,YL,O,O,
O,O,BLUE,O,O,YL,O,O,
O,O,BLUE,O,O,YL,O,O,
O,O,BLUE,YL,O,YL,O,YL,
O,O,O,O,YL,YL,YL,O,
O,O,O,O,O,YL,O,O
]

# Define event callbacks
def on_connect(client, userdata, flags, rc):
    print("rc: " + str(rc))

def on_message(client, obj, msg):
    print(msg.topic + " " + str(msg.qos) + " " + str(msg.payload))

def on_publish(client, obj, mid):
    print("mid: " + str(mid))

def on_subscribe(client, obj, mid, granted_qos):
    print("Subscribed: " + str(mid) + " " + str(granted_qos))
    if granted_qos == "":
        print("not info")
    else:
        status = json.loads(grandted_qos)
        print(status)
    

def on_log(client, obj, level, string):
    print(string)
    
mqttc = mqtt.Client()
mqttc.on_message = on_message
mqttc.on_connect = on_connect
mqttc.on_subscribe = on_subscribe
# Uncomment to enable debug messages
#mqttc.on_log = on_log

# Parse CLOUDMQTT_URL (or fallback to localhost)
url_str = os.environ.get('CLOUDMQTT_URL', "mqtt://hrrgcrqx:af3wqGskmMfY@m20.cloudmqtt.com:12771")
url = urlparse.urlparse(url_str)

# Connect
mqttc.username_pw_set(url.username, url.password)
mqttc.connect(url.hostname, url.port)
mqttc.subscribe("picture/test", 0)
# Publish a message
rc = 0
while rc == 0:
    rc = mqttc.loop()
    print("rc: " + str(rc))

Application/app/src/main/java/imerir/scavengerhunt/SendPicture.java
package imerir.scavengerhunt;

import android.Manifest;
import android.app.ListActivity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.media.ExifInterface;
import android.support.v7.app.AppCompatActivity;
import android.util.Base64;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.gms.vision.CameraSource;
import com.google.android.gms.vision.barcode.BarcodeDetector;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;

import static imerir.scavengerhunt.MainActivity.PREFS;
import static imerir.scavengerhunt.MainActivity.TAG;
import static imerir.scavengerhunt.MainActivity.contURl;
import static java.lang.String.format;

public class SendPicture extends AppCompatActivity {

    private static final int FILE_SELECT_CODE = 0;
    TextView mDistance;
    Button select;
    Button sendPicture;
    Button back;
    ImageView image;
    RequestQueue mRequestQueue;
    File imgFile;
    infoPeriph infoPeri;
    boolean pictureSelected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_send_picture);
        select = findViewById(R.id.selectpicture);
        sendPicture = findViewById(R.id.button2);
        mDistance = findViewById(R.id.dist);
        image = findViewById(R.id.imageView);
        back = findViewById(R.id.backButton);

        mRequestQueue = Volley.newRequestQueue(this);

        infoPeri = infoPeriph.getInstance();

        pictureSelected = false;

        sendPicture.setEnabled(false);

        sendPicture.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                Toast toast = Toast.makeText(SendPicture.this, "Envoie de la photo, merci de patienter ", Toast.LENGTH_LONG);
                toast.show();
                httpPostImage(contURl+"/picture");

            }
        });

        select.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                showFileChooser();

            }
        });
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                switchPager();

            }
        });

        Thread t = new Thread() {

            @Override
            public void run() {
                try {
                    while (!isInterrupted()) {
                        Thread.sleep(1000);
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                mDistance.setText(format("%.2f", infoPeri.getDistance()) + " m");
                                if (infoPeri.getDistance()<3.0){
                                    mDistance.setTextColor(Color.GREEN);
                                    if (pictureSelected){
                                        sendPicture.setEnabled(true);
                                    }
                                    else{
                                        sendPicture.setEnabled(false);
                                    }
                                }
                                else if (infoPeri.getDistance() == 50.0){
                                    mDistance.setText("");
                                }
                                else{
                                    mDistance.setTextColor(Color.RED);
                                    sendPicture.setEnabled(false);

                                }
                            }
                        });
                    }
                } catch (InterruptedException e) {
                    Log.e(TAG,e.getMessage());
                }
            }
        };

        t.start();
    }

    private void showFileChooser() {

        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("*/*");      //all files
        intent.addCategory(Intent.CATEGORY_OPENABLE);

        try {
            startActivityForResult(Intent.createChooser(intent, "Select a File to Upload"), FILE_SELECT_CODE);
        } catch (android.content.ActivityNotFoundException ex) {
            // Potentially direct the user to the Market with a Dialog
            Toast.makeText(SendPicture.this, "Please install a File Manager.", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case FILE_SELECT_CODE:
                if (resultCode == RESULT_OK) {
                    // Get the Uri of the selected file
                    Uri uri = data.getData();
                    Log.i(TAG, "File Uri: " + uri.toString());
                    // Get the path
                    String path = null;
                    try {
                        path = getPath(this, uri);
                    } catch (URISyntaxException e) {
                        Log.e(TAG, e.getMessage());                    }
                    Log.d(TAG, "File Path: " + path);

                    imgFile = new File(path);

                    if(imgFile.exists()){

                        Bitmap myBitmap = BitmapFactory.decodeFile(imgFile.getAbsolutePath());

                        image.setImageBitmap(myBitmap);
                        pictureSelected = true;

                    }
                }
                break;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    void httpPostImage(String url) {

        SharedPreferences prefs = getSharedPreferences(PREFS, MODE_PRIVATE);
        String deviceID = prefs.getString("DeviceID", null);

        String latitute = "";
        String longitude = "";
        try {
            final ExifInterface exifInterface = new ExifInterface(imgFile.getAbsolutePath());
            latitute = exifInterface.getAttribute(ExifInterface.TAG_GPS_LATITUDE);
            longitude = exifInterface.getAttribute(ExifInterface.TAG_GPS_LONGITUDE);

        } catch (IOException e) {
            Log.e("TAG", e.getMessage());
        }

        String image64 = jpegTo64();

        String[] splitString = imgFile.getName().split("_");

        JSONObject postData = new JSONObject();
        try {
            postData.put("id_Equipe", deviceID);
            postData.put("name",splitString[0]);
            postData.put("latitute", latitute);
            postData.put("longitude", longitude);
            postData.put("image", image64);
        } catch (Exception e) {
            Log.e(TAG,e.getMessage());
        }

        Response.Listener<JSONObject> onSuccess = new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {
                    Toast toast = Toast.makeText(SendPicture.this, "Photo analysée ! Vérifiez l'afficheur !", Toast.LENGTH_LONG);
                    toast.show();
                } catch (Exception e) {
                    Toast toast = Toast.makeText(SendPicture.this, "Erreur de lecture de l'image", Toast.LENGTH_LONG);
                    toast.show();
                }
            }
        };

        Response.ErrorListener onError = new Response.ErrorListener() {
            public void onErrorResponse(VolleyError error) {
                Toast toast = Toast.makeText(SendPicture.this, "Image incorrect", Toast.LENGTH_LONG);
                toast.show();
                Log.i("Message Recu :","Erreur lors de la requête");
            }
        };



        JsonObjectRequest request = new JsonObjectRequest(Request.Method.POST, url, postData, onSuccess, onError);
        request.setRetryPolicy(new DefaultRetryPolicy(10000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        mRequestQueue.add(request);
    }
    public static String getPath(Context context, Uri uri) throws URISyntaxException {
        if ("content".equalsIgnoreCase(uri.getScheme())) {
            String[] projection = { "_data" };
            Cursor cursor;

            try {
                cursor = context.getContentResolver().query(uri, projection, null, null, null);
                int column_index = cursor.getColumnIndexOrThrow("_data");
                if (cursor.moveToFirst()) {
                    return cursor.getString(column_index);
                }
            } catch (Exception e) {
                Log.e(TAG,e.getMessage());
            }
        }
        else if ("file".equalsIgnoreCase(uri.getScheme())) {
            return uri.getPath();
        }

        return null;
    }
    public String jpegTo64(){
        Bitmap myBitmap = Bitmap.createScaledBitmap(BitmapFactory.decodeFile(imgFile.getAbsolutePath()),200,200,true);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        myBitmap.compress(Bitmap.CompressFormat.JPEG, 20, baos);
        byte[] imageBytes = baos.toByteArray();
        return Base64.encodeToString(imageBytes, Base64.DEFAULT);
    }

    void switchPager() {
        finish();
        Intent i = new Intent(SendPicture.this, ListeActivity.class);
        SendPicture.this.startActivity(i);
    }

}


Application/app/src/main/java/imerir/scavengerhunt/MainActivity.java
package imerir.scavengerhunt;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.RemoteException;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.util.SparseArray;
import android.widget.TextView;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.gms.vision.CameraSource;
import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.barcode.Barcode;
import com.google.android.gms.vision.barcode.BarcodeDetector;

import org.altbeacon.beacon.Beacon;
import org.altbeacon.beacon.BeaconConsumer;
import org.altbeacon.beacon.BeaconManager;
import org.altbeacon.beacon.BeaconParser;
import org.altbeacon.beacon.Identifier;
import org.altbeacon.beacon.MonitorNotifier;
import org.altbeacon.beacon.RangeNotifier;
import org.altbeacon.beacon.Region;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Collection;

import static java.lang.String.format;


public class MainActivity extends AppCompatActivity implements BeaconConsumer {

    private static final int REQUEST_PERMISSION = 1;
    protected static final String TAG = "Scavenger Hunt";
    static final int REQUEST_PERMCAM = 1;
    public static final String PREFS = "PREFS";
    SharedPreferences sharedPreferences;
    public static final String contURl = "http://51.254.121.94:4000";
    //public static final String contURl = "https://routerint.mignolet.fr";
    //public static final String contURl = "http://172.30.1.35:5003";

    String id = null;
    String id2 = null;
    Region region1;
    Region region2;
    infoPeriph infoPeri;
    boolean inscrit = false;
    int typeRegion = 1;
    double distSendBeacon;

    private BeaconManager beaconManager;
    RequestQueue mRequestQueue;

    TextView mTextView;
    TextView mDistance;


    void displayText(final String message) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mTextView.setText(message);
            }
        });
    }

    // Model Components
    public static BarcodeDetector mDetector;
    public static CameraSource mCamera;


    @Override
    protected void onDestroy() {
        super.onDestroy();
        mCamera.stop();
        mDetector.release();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        infoPeri = infoPeriph.getInstance();

        mRequestQueue = Volley.newRequestQueue(this);
        sharedPreferences = getBaseContext().getSharedPreferences(PREFS, MODE_PRIVATE);

        mTextView = findViewById(R.id.textView);
        mDistance = findViewById(R.id.distance);

        postHttpRequest(contURl+"/team");

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
            if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(new String[]{Manifest.permission.CAMERA}, REQUEST_PERMCAM);
                setupDetectorAndCamera();
            }
            if (checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                showAlertDialog("This app needs location access",
                        "Please grant location access so this app can detect beacons.", new DialogInterface.OnDismissListener() {
                            @Override
                            public void onDismiss(DialogInterface dialog) {
                                requestPermissions(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, REQUEST_PERMISSION);
                            }
                        });
            }
            if (checkSelfPermission(Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
                showAlertDialog("This app needs location access",
                        "Please grant location access so this app can detect beacons.", new DialogInterface.OnDismissListener() {
                            @Override
                            public void onDismiss(DialogInterface dialog) {
                                requestPermissions(new String[]{Manifest.permission.READ_PHONE_STATE}, REQUEST_PERMISSION);
                            }
                        });
            }
        } else {
            setupDetectorAndCamera();
        }
    }


    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case REQUEST_PERMISSION: {
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    Log.d(TAG, "coarse location permission granted");
                } else {
                    showAlertDialog("Functionality limited",
                            "Since location access has not been granted, this app will not be able to discover beacons when in the background.", null);
                }
                return;
            }
        }
        if (requestCode == REQUEST_PERMCAM) {
            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                setupDetectorAndCamera(); // continue the setup
            } else {
                Log.e(TAG, "Permission was denied or request was cancelled.");
                displayText("Access to the camera is REQUIRED.");
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        beaconManager = BeaconManager.getInstanceForApplication(this);

        final String iBeaconLayout = "m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24";
        beaconManager.getBeaconParsers().add(new BeaconParser().setBeaconLayout(iBeaconLayout));

        beaconManager.bind(this);
        postHttpRequest(contURl+"/team");

    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    private void showAlertDialog(String title, String message, DialogInterface.OnDismissListener onDismissListener) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title);
        builder.setMessage(message);
        builder.setPositiveButton(android.R.string.ok, null);
        if (onDismissListener != null) builder.setOnDismissListener(onDismissListener);
        builder.show();
    }

    @Override
    public void onBeaconServiceConnect() {

        Log.i(TAG, "beaconServiceConnected");
        beaconManager.addRangeNotifier(new RangeNotifier() {

            @Override
            public void didRangeBeaconsInRegion(Collection<Beacon> collection, Region region) {
                for (Beacon beacon : collection) {
                    Log.i(TAG, "Detected beacon : " + beacon.getId1());
                    Log.i(TAG, "Detected beacon @ distance " + beacon.getDistance());
                    if (typeRegion == 1){
                        final double dist = beacon.getDistance();
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                mDistance.setTextColor(Color.RED);
                                mDistance.setText(format("%.2f", dist) + " m");

                            }
                        });
                        if (inscrit){
                            getHttpRequest(contURl+"/beaconSendPicture",2);
                            switchPager();
                        }
                        else{
                            if (dist < 3.0){
                                postHttpRequest(contURl+"/inscript");
                                getHttpRequest(contURl+"/beaconSendPicture",2);
                                postHttpRequest(contURl+"/team");
                                switchPager();
                            }
                        }
                    }
                    else{
                        distSendBeacon = beacon.getDistance();
                        infoPeri.setDistance(distSendBeacon);

                    }


                }
            }
        });
        beaconManager.addMonitorNotifier(new MonitorNotifier() {
            @Override
            public void didEnterRegion(Region region) {
                Log.i(TAG, "didEnterRegion");
            }

            @Override
            public void didExitRegion(Region region) {
                Log.i(TAG, "didExitRegion");
            }

            @Override
            public void didDetermineStateForRegion(int i, Region region) {
                Log.i(TAG, "didDetermineStateForRegion = " + i);
            }
        });
        getHttpRequest(contURl+"/beaconInscript",1);

    }

    void setupDetectorAndCamera() {
        // build the detector
        mDetector = new BarcodeDetector.Builder(getApplicationContext())
                .setBarcodeFormats(Barcode.QR_CODE)
                .build();

        mDetector.setProcessor(new Detector.Processor<Barcode>() {
            @Override
            public void release() {
            }

            @Override
            public void receiveDetections(Detector.Detections<Barcode> detections) {
                final SparseArray<Barcode> barcodes = detections.getDetectedItems();
                processBarcodes(barcodes);
            }
        });

        if (!mDetector.isOperational()) {
            mTextView.setText("Could not set up the detector!\nPlease update or upgrade your tablet (consider an iPhone X).");
            return;
        }

        // build the camera
        CameraSource.Builder cameraBuilder = new CameraSource.Builder(this, mDetector);
        cameraBuilder.setAutoFocusEnabled(true);
        cameraBuilder.setFacing(CameraSource.CAMERA_FACING_BACK);
        cameraBuilder.setRequestedFps(10);

        mCamera = cameraBuilder.build();

        // start the camera
        try {
            mCamera.start();
        } catch (SecurityException e) { // camera not allowed
            Log.e(TAG, e.getMessage());
            mTextView.setText("Failed to start the camera");
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            mTextView.setText("Failed to start the camera");
        }
    }

    void processBarcodes(SparseArray<Barcode> barcodes) {
        if (barcodes.size() != 0) {
            String url = barcodes.valueAt(0).displayValue;
            displayText(url);
            if (inscrit){
                getHttpRequest(contURl+"/beaconSendPicture",2);
                switchPager();
            }
            else{

                postHttpRequest(contURl+"/inscript");
                getHttpRequest(contURl+"/beaconSendPicture",2);
                postHttpRequest(contURl+"/team");
                switchPager();

            }
        } else {
            displayText("No barcode detected.");
        }
    }

    void getHttpRequest(String url,final int requestType) {
        Log.i(TAG,url);

        Response.Listener<JSONObject> onSuccess = new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                Log.i(TAG, response.toString());
                try {
                    if (requestType == 1){
                        typeRegion =1;
                        id = response.getJSONObject("message").getString("uid");
                        region1 = new Region("Totorama", Identifier.parse(id), null, null);
                        beaconManager.startMonitoringBeaconsInRegion(region1);
                        beaconManager.startRangingBeaconsInRegion(region1);
                    }
                    if (requestType == 2){
                        typeRegion = 2;
                        id2 = response.getJSONObject("message").getString("uid");
                        region2 = new Region("Totorama", Identifier.parse(id), null, null);
                        beaconManager.stopMonitoringBeaconsInRegion(region1);
                        beaconManager.stopRangingBeaconsInRegion(region1);
                        beaconManager.startMonitoringBeaconsInRegion(region2);
                        beaconManager.startRangingBeaconsInRegion(region2);
                    }


                } catch (JSONException e) {
                    Log.e(TAG, e.getMessage());
                } catch (RemoteException e) {
                    Log.e(TAG, e.getMessage());
                }
            }

        };

        Response.ErrorListener onError = new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                try{
                    Log.e(TAG,error.getMessage());
                    Log.i(TAG,"Erreur lors de la requête");
                }catch (NullPointerException e){
                    Log.e(TAG,e.getMessage());
                }

            }
        };

        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null, onSuccess, onError);

        mRequestQueue.add(request);

    }

    void postHttpRequest(final String url) {
        Log.i(TAG,url);
        JSONObject postData = new JSONObject();
        try {
            postData.put("id", getDeviceId(this));
        } catch (Exception e) {
            Log.e(TAG,e.getMessage());
        }

        Response.Listener<JSONObject> onSuccess = new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {
                    Log.i(TAG,response.toString());
                    if (url.contains("team")){
                        inscrit = true;
                    }

                } catch (Exception e) {
                    Log.e(TAG,e.getMessage());
                }
            }
        };

        Response.ErrorListener onError = new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {

                try{
                    Log.e(TAG,error.getMessage());
                    Log.i(TAG,"Erreur lors de la requête");
                }catch (NullPointerException e){
                    Log.e(TAG,e.getMessage());
                }
            }
        };

        JsonObjectRequest request = new JsonObjectRequest(Request.Method.POST, url, postData, onSuccess, onError);

        mRequestQueue.add(request);
    }

    public String getDeviceId(Context context) {
        String androidId = Settings.Secure.getString(context.getContentResolver(),Settings.Secure.ANDROID_ID) + Build.SERIAL;
        Log.d(TAG,androidId);

        sharedPreferences.edit().putString("DeviceID",androidId).apply();
        return androidId;

    }

    void switchPager(){
        Intent i = new Intent(MainActivity.this, ListeActivity.class);
        MainActivity.this.startActivity(i);

    }

}
